### PR TITLE
Fixes bad camera positions and change options WMTS

### PR DESCRIPTION
**Pull request by gchoqueux**
Fixes camera bad positions when you use :

* setCameraTargetGeoPosition() #255
* setRange()
* pan() see #260
* createSceneGlobe()

Move WMTS's options in layer.options 

Add little doc addImageryLayer and addElevationLayer

**Warning, this PR doesn't fix incorrect position with setZoomLevel and setZoomScale**

### CODE REVIEW

**Comment by peppsac** [comment]
Please move the test commit out of this PR (big topic so it deserves its own PR).

**Comment by nosy-b** [comment]
I just started reviewing with method` ApiGlobe.getCenter()`
In index.html it seems the initial position of camera is never properly used but is rounded up. 
Try for example to start centerd on Paris` const positionOnGlobe = { longitude: 2.351323, latitude: 48.856712, altitude: 25000000 };` 
Then ` ApiGlobe.getCenter()` will return {longitude: 2.3513230022207057, latitude: 49.00976271512532} which seems to be the real position as zooming get you north of Paris.

File: src/Core/Commander/Providers/WMTS_Provider.js [hunk 1]
```diff
@@ -67,30 +67,32 @@ WMTS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer
...
+            const arrayLimits = Object.keys(options.tileMatrixSetLimits);
+            const size = arrayLimits.length;
+            const maxZoom = Number(arrayLimits[size - 1]);
+            const minZoom = maxZoom - size + 1;
 
-        layer.zoom = {
-            min: minZoom,
-            max: maxZoom,
-        };
+            layer.zoom = {
```
Comment (elias75015) [inline src/Core/Commander/Providers/WMTS_Provider.js]: Shouldn't it be in "layer.options" ?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 2]
```diff
@@ -634,15 +642,26 @@ ApiGlobe.prototype.setCenter = function setCenter(coordinates, isAnimated) {
...
 ApiGlobe.prototype.setCenterAdvanced = function setCenterAdvanced(pPosition, isAnimated) {
     isAnimated = isAnimated || this.isAnimationEnabled();
     return this.setCenter(pPosition, isAnimated).then(() => {
-        const p = this.scene.currentControls().setOrbitalPosition(undefined, pPosition.heading, pPosition.tilt, isAnimated);
-        return p;
+        if (pPosition.range === undefined) {
+            if (pPosition.level) {
+                pPosition.range = this.getRangeFromZoomLevel(pPosition.level);
+            } else if (pPosition.scale) {
+                pPosition.range = this.getRangeFromScale(pPosition.scale);
```
Comment (elias75015) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: - Only with the param "isAnimated" set to false : When the scale parameter is given, the view is tilted to 90°, even without the parameter "tilt" given to the function.
- I think it is more explicit to give the scale param as a fraction denominator than in decimal (ex : 500 instead of 0.002)

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 3]
```diff
@@ -634,15 +642,26 @@ ApiGlobe.prototype.setCenter = function setCenter(coordinates, isAnimated) {
...
+ * @return {Promise}
  */
 ApiGlobe.prototype.setCenterAdvanced = function setCenterAdvanced(pPosition, isAnimated) {
     isAnimated = isAnimated || this.isAnimationEnabled();
     return this.setCenter(pPosition, isAnimated).then(() => {
-        const p = this.scene.currentControls().setOrbitalPosition(undefined, pPosition.heading, pPosition.tilt, isAnimated);
-        return p;
+        if (pPosition.range === undefined) {
+            if (pPosition.level) {
+                pPosition.range = this.getRangeFromZoomLevel(pPosition.level);
```
Comment (elias75015) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: The function takes well into account the level param (the view seems to be correctly positioned), but the getZoomLevel doesn't return the same value as the one given in the setCenterAdvenced function.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 4]
```diff
@@ -428,7 +428,9 @@ ApiGlobe.prototype.getCameraLocation = function getCameraLocation() {
 
 ApiGlobe.prototype.getCenter = function getCenter() {
     var controlCam = this.scene.currentControls();
-    return this.projection.cartesianToGeo(controlCam.getTargetCameraPosition());
+    const tg = controlCam.getTargetCameraPosition();
+    const coordinates = new Coordinates('EPSG:4978', tg.x, tg.y, tg.z).as('EPSG:4326');
+    return { longitude: coordinates.longitude(UNIT.DEGREE), latitude: coordinates.latitude(UNIT.DEGREE) };
```
Comment (elias75015) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Not sure about the length of the returned values (does it make sense ?) : 
latitude : 47.995276640947594
longitude : 1.0015039089052769

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 5]
```diff
@@ -660,7 +679,7 @@ ApiGlobe.prototype.setRange = function setRange(pRange, isAnimated) {
 
     return this.scene.currentControls().setRange(pRange, isAnimated).then(() => {
         this.scene.notifyChange(1);
-        this.setSceneLoaded().then(() => {
+        return this.setSceneLoaded().then(() => {
             this.scene.currentControls().updateCameraTransformation();
             this.viewerDiv.dispatchEvent(eventRange);
```
Comment (elias75015) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:735]: If I do a setRange(10000), the view position is OK, but the getRange() does not return 10000.

The getRange seems to return old range value. Ex :
- setRange(10000) --> getRange() = 25011386.530042846
- setRange(200000) --> getRange() =10013.066997543165 (almost 10000)

**Comment by gchoqueux** [comment]
> Then ApiGlobe.getCenter() will return {longitude: 2.3513230022207057, latitude: 49.00976271512532} which seems to be the real position as zooming get you north of Paris.

Fix in last commit, It was due to an ellipsoidal/sphere approximation

File: src/Core/Commander/Providers/WMTS_Provider.js (hunk 1 above)
Comment (gchoqueux) [reply src/Core/Commander/Providers/WMTS_Provider.js]: ok, I move in options

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 4 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Yes maybe, especially since I do not know the precision of my calculation

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 3 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: getZoomLevel will be fixed in #279

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 2 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: > I think it is more explicit to give the scale param as a fraction denominator than in decimal (ex : 500 instead of 0.002)

The scale of a map is the ratio,  we must be find a another name to this parameter

Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: > Only with the param "isAnimated" set to false : When the scale parameter is given, the view is tilted to 90°, even without the parameter "tilt" given to the function.

Try with the new commit ` Fix (Viewer) :  add default value for setCenterAdvanced() `

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 5 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:735]: try with `then` promise,  you do getRange() just after setRange(), While the setRange is not yet completed 

setRange(10000).then(() => getRange())

**Comment by nosy-b** [comment]
LGTM

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 6]
```diff
@@ -260,13 +303,13 @@ ApiGlobe.prototype.addElevationLayersFromJSONArray = function addElevationLayers
 ApiGlobe.prototype.getMinZoomLevel = function getMinZoomLevel(index) {
     var layer = this.getImageryLayers()[index];
     if (layer && layer.zoom) {
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: I guess it should be `layer.options.zoom`.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 7]
```diff
@@ -260,13 +303,13 @@ ApiGlobe.prototype.addElevationLayersFromJSONArray = function addElevationLayers
...
-        return layer.zoom.min;
+        return layer.options.zoom.min;
     } else {
         var layers = this.getImageryLayers();
         let min = Infinity;
         for (var i = layers.length - 1; i >= 0; i--) {
             if (layers[i].zoom) {
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: same here.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 8]
```diff
@@ -628,7 +672,14 @@ ApiGlobe.prototype.setCameraTargetGeoPosition = function setCameraTargetGeoPosit
...
         .as('EPSG:4978').xyz();
-    position3D.range = coordinates.range;
+    if (coordinates.range) {
+        // error delta ellipsoid
+        const currentTargetPosition = this.getCameraTargetGeoPosition();
+        const oldPosition3D = new C.EPSG_4326(currentTargetPosition.longitude, currentTargetPosition.latitude, 0)
+            .as('EPSG:4978').xyz();
+        const diff = position3D.length() - oldPosition3D.length();
+        position3D.range = coordinates.range + diff;
+    }
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Suggestion: put this piece of code in GlobeControls.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 9]
```diff
@@ -643,15 +694,37 @@ ApiGlobe.prototype.setCameraTargetGeoPosition = function setCameraTargetGeoPosit
...
  * <iframe width="100%" height="400" src="//jsfiddle.net/iTownsIGN/7yk0mpn0/embedded/" allowfullscreen="allowfullscreen" frameborder="0"></iframe>
- * @param {Position} pPosition - The detailed position in the scene.
- * @param      {boolean}  isAnimated  Indicates if animated
- * @return     {Promise}
- */
-ApiGlobe.prototype.setCameraTargetGeoPositionAdvanced = function setCameraTargetGeoPositionAdvanced(pPosition, isAnimated) {
+ * @param {Position} position
+ * @param {Number}  position.longitude  Coordinate longitude WGS84 in degree
+ * @param {Number}  position.latitude  Coordinate latitude WGS84 in degree
+ * @param {Number}  [position.tilt]  Camera tilt in degree
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: should not be surrounded by `[]`, right?

**Review by autra — COMMENTED** [review]
Leaving some comments, as there are still some `layer.zoom` in the code. You might want to use `git grep` to make sure nothing is left.

Btw, why not keeping `layer.zoom` and not using `layer.options.zoom`? Do we really need the additional level (as more or less everything are layer options anyway)? 

Thanks

**Review by peppsac — DISMISSED** [review]
Regardless of the changes themselves (no time for review atm) this PR has several problems:
- no PR summary (and the title is vague)
- 7 commits with broad topics
- commits messages are either empty or meaningless (e.g : `Now setRange() return Promise`)

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 9 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: position.tilt is optionnal

**Comment by gchoqueux** [comment]
> layer.options.zoom

zoom is specific to WMTS, so I moved zoom in `wmts_options` (see doc)

but this is not perfect, `attribution` should out of options

File: src/Core/Commander/Providers/WMTS_Provider.js [hunk 10]
```diff
@@ -241,9 +243,9 @@ WMTS_Provider.prototype.computeLevelToDownload = function computeLevelToDownload
     var lvl = ancestor ? ancestor.level : tile.level;
 
     return Math.min(
-        layer.zoom.max,
+        layer.options.zoom.max,
```
Comment (elias75015) [inline src/Core/Commander/Providers/WMTS_Provider.js:244]: A lot of 404 CORS errors are launched errors when we zoom beyond the zoom.max given as parameter of the layer.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 11]
```diff
@@ -643,15 +694,37 @@ ApiGlobe.prototype.setCameraTargetGeoPosition = function setCameraTargetGeoPosit
...
+ * @param {Number}  [position.heading]  Camera heading in degree
+ * @param {Number}  [position.range]  The camera distance to the target center
+ * @param {Number}  [position.level]  level,  ignored if range is set
+ * @param {Number}  [position.scale]  scale,  ignored if the zoom level or range is set. For a scale of 1/500 it is necessary to write 0,002.
+ * @param {boolean}  isAnimated  Indicates if animated
+ * @return {Promise}
+ */
+ApiGlobe.prototype.setCameraTargetGeoPositionAdvanced = function setCameraTargetGeoPositionAdvanced(position, isAnimated) {
```
Comment (elias75015) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Not totally working for the moment (tested in web dev console) :
- range parameter is strangely handled (value given doesn't fit with the view)
- heading and tilt parameters not taken into account

**Comment by gchoqueux** [comment]
@autra @peppsac  : remarks take into account

**Comment by gchoqueux** [comment]
@elias75015 Branch is updated

**Comment by gchoqueux** [comment]
@peppsac when will you be available?

**Comment by peppsac** [comment]
@gchoqueux not this week, maybe next one.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 9 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: I agree with @gchoqueux

Comment (autra) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Yes yes, me too :-) Thanks for the precision

**Comment by gchoqueux** [comment]
@autra @peppsac, this PR needs reviewing

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 12]
```diff
@@ -16,7 +16,8 @@ import Fetcher from '../../Providers/Fetcher';
 import { STRATEGY_MIN_NETWORK_TRAFFIC } from '../../../../Scene/LayerUpdateStrategy';
 
 var sceneIsLoaded = false;
-var eventLoaded = new CustomEvent('globe-loaded');
+var eventLoaded = new CustomEvent('dataloaded');
+var eventInitialized = new CustomEvent('initialized');
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Create events where you dispatch them. They can't be used several time safely

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 13]
```diff
@@ -97,10 +97,49 @@ ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer) {
...
+ * @property {Object} [zoom.min] layer's zoom minimum
+ * @property {Object} [zoom.max] layer's zoom maximum
+ */
+
+/**
+ * This function adds an imagery layer to the scene. The layer id must be unique.
+ * The protocol rules wich parameters are then needed for the function.
  * @constructor
- * @param {Layer} layer.
+ * @param {Layer} layer
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Missing the `Layer` typedef

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 14]
```diff
@@ -97,10 +97,49 @@ ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer) {
...
- * This function adds an imagery layer to the scene. The layer id must be unique. The protocol rules wich parameters are then needed for the function.
+ * The intellectual property rights
+ * @typedef {Object} Attribution
+ * @property {String} name
+ * @property {String} url
+ */
+
+/**
+ * Options to wms protocol
+ * @typedef {Object} options_wms
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: By tradition, types are in CamelCase (and variables are also camelCase in itowns anyway).

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 15]
```diff
@@ -97,10 +97,49 @@ ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer) {
 };
 
 /**
- * This function adds an imagery layer to the scene. The layer id must be unique. The protocol rules wich parameters are then needed for the function.
+ * The intellectual property rights
+ * @typedef {Object} Attribution
+ * @property {String} name
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: the type should be the same string returned by the `typeof` operator, so `string` and not `String`

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 16]
```diff
@@ -97,10 +97,49 @@ ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer) {
...
+
+/**
+ * Options to wtms protocol
+ * @typedef {Object} options_wmts
+ * @property {Attribution} attribution The intellectual property rights for the layer
+ * @property {String} name
+ * @property {String} mimetype
+ * @property {String} tileMatrixSet
+ * @property {Array.<Object>} tileMatrixSetLimits The limits for the tile matrix set
+ * @property {Number} tileMatrixSetLimits.minTileRow Minimum row for tiles at the level
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: `number`

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 12 above)
Comment (autra) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Actually, for all the events, but particularly for the events you modify in this PR: the only things it makes sense to put in `const` (not `var` btw) are the string itself:

```js
const DATA_LOADED_EVENT = 'dataloaded';

/* later */

this.dispatchEvent(new CustomEvent(DATA_LOADED_EVENT));

/* elsewhere */

itowns.viewer.addEventListener(itowns.viewer.DATA_LOADED_EVENT, /* ... */);
```

Webpack will then warn if there is a typo.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 17]
```diff
@@ -643,15 +689,37 @@ ApiGlobe.prototype.setCameraTargetGeoPosition = function setCameraTargetGeoPosit
...
+ * @param {boolean}  isAnimated  Indicates if animated
+ * @return {Promise}
+ */
+ApiGlobe.prototype.setCameraTargetGeoPositionAdvanced = function setCameraTargetGeoPositionAdvanced(position, isAnimated) {
     isAnimated = isAnimated || this.isAnimationEnabled();
-    return this.setCameraTargetGeoPosition(pPosition, isAnimated).then(() => {
-        const p = this.scene.currentControls().setOrbitalPosition(undefined, pPosition.heading, pPosition.tilt, isAnimated);
-        return p;
+    return this.setCameraTargetGeoPosition(position, isAnimated).then(() => {
+        if (position.range === undefined) {
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: `if (position.range)`

File: src/Renderer/ThreeExtended/GlobeControls.js [hunk 18]
```diff
@@ -478,18 +478,27 @@ function GlobeControls(camera, domElement, engine) {
...
 
+            if (enableTargetHelper) {
+                pickingHelper.visible = true;
+                cameraTargetOnGlobe.visible = true;
+            }
+
             return position;
         };
-    }());
+    }(this.pickingHelper));
```
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js]: Why do you need this closure here? `this.pickingHelper` won't change, I don't see the point of capturing engineGFX, and position is not really an internal state here, as you never read from it.

File: src/Renderer/ThreeExtended/GlobeControls.js [hunk 19]
```diff
@@ -478,18 +478,27 @@ function GlobeControls(camera, domElement, engine) {
...
 
-        return function getPickingPosition(coords)
-        {
             position = engineGfx.getPickingPositionFromDepth(coords);
             engineGfx.renderScene();
 
+            if (enableTargetHelper) {
+                pickingHelper.visible = true;
+                cameraTargetOnGlobe.visible = true;
+            }
```
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js]: Can you explain why you need these 2 `if` here? It looks odd to have to do that.

File: src/Renderer/ThreeExtended/GlobeControls.js [hunk 20]
```diff
@@ -669,6 +679,9 @@ function GlobeControls(camera, domElement, engine) {
         spherical.setFromVector3(offset);
         state = CONTROL_STATE.NONE;
         lastRotation = [];
+        if (enableTargetHelper) {
+            this.dispatchEvent(this.changeEvent);
```
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js:679]: Note: typically why I'm pretty sure the dependency between c3DEngine and Controls should not exist.

File: src/Renderer/ThreeExtended/GlobeControls.js [hunk 21]
```diff
@@ -1122,13 +1135,15 @@ function GlobeControls(camera, domElement, engine) {
...
     // update object camera position
-    this.updateCameraTransformation = function updateCameraTransformation(controlState)
+    this.updateCameraTransformation = function updateCameraTransformation(controlState, forceNoUpdateTarget)
     {
         const bkDamping = this.enableDamping;
         this.enableDamping = false;
         state = controlState || CONTROL_STATE.ORBIT;
         update();
-        updateCameraTargetOnGlobe.bind(this)();
+        if (forceNoUpdateTarget === undefined || forceNoUpdateTarget) {
```
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js]: Please use positive logic : `forceUpdate`, `preventUpdate`, whatever you prefer. Negative naming is difficult to read and error-prone. Indeed, here when forceNoUpdateTarget is true, we *do* update the target on globe, which seems contradictory to me.

Btw, why do you need to bypass the update of the camera in some case ?

File: src/Renderer/ThreeExtended/GlobeControls.js [hunk 22]
```diff
@@ -1269,6 +1283,15 @@ GlobeControls.prototype.getCameraTargetPosition = function getCameraTargetPositi
 GlobeControls.prototype.setCameraTargetPosition = function setCameraTargetPosition(position, isAnimated) {
     const center = this.getCameraTargetPosition();
 
+    if (position.range) {
+        // Compensation of the altitude from the approximation of the ellipsoid by a sphere
```
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js:1283]: Then the question is: where do we approximate the ellipsoid by a sphere that in turn make this hack necessary?

File: src/Renderer/c3DEngine.js [hunk 23]
```diff
@@ -15,7 +15,7 @@ import RendererConstant from './RendererConstant';
 
 var instance3DEngine = null;
 
-function c3DEngine(scene, positionCamera, viewerDiv, debugMode, gLDebug) {
+function c3DEngine(scene, positionCamera, targetCamera, viewerDiv, debugMode, gLDebug) {
```
Comment (autra) [inline src/Renderer/c3DEngine.js]: This is specific to GlobeControls, so to a Globe. I would prefer not to add that as a parameter to c3DEngine.

File: src/Scene/Scene.js [hunk 24]
```diff
@@ -39,12 +39,17 @@ const RENDERING_ACTIVE = 1;
  * - remove debug boolean, replace by if __DEBUG__ and checkboxes in debug UI
  * - Scene (and subobjects) should be instanciable several times.
  */
-function Scene(crs, positionCamera, viewerDiv, debugMode, gLDebug) {
+function Scene(crs, coordinate, viewerDiv, debugMode, gLDebug) {
```
Comment (autra) [inline src/Scene/Scene.js]: Why this change? you don't need it and it's less explicit for the caller.

File: src/Scene/Scene.js [hunk 25]
```diff
@@ -39,12 +39,17 @@ const RENDERING_ACTIVE = 1;
...
     if (instanceScene !== null) {
         throw new Error('Cannot instantiate more than one Scene');
     }
     this.referenceCrs = crs;
 
+    const positionCamera = coordinate.as(crs);
+    const coordinateTargetCamera = coordinate.clone();
+    coordinateTargetCamera._values[2] = 0;
```
Comment (autra) [inline src/Scene/Scene.js]: undescored properties are meant to be considered as private. You can't use it here. You either need to move this code or to add a function to Coordinate (`fromLonLatHeight` maybe)

**Review by autra — CHANGES_REQUESTED** [review]

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 17 above)
Comment (gchoqueux) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: It is necessary to distinguish range equal zero and undefined

File: src/Renderer/ThreeExtended/GlobeControls.js (hunk 19 above)
Comment (gchoqueux) [reply src/Renderer/ThreeExtended/GlobeControls.js]: if, there are targets helper, we must hide these targets for the picking

File: src/Renderer/ThreeExtended/GlobeControls.js (hunk 18 above)
Comment (gchoqueux) [reply src/Renderer/ThreeExtended/GlobeControls.js]: Yes indeed

File: src/Renderer/ThreeExtended/GlobeControls.js (hunk 20 above)
Comment (gchoqueux) [reply src/Renderer/ThreeExtended/GlobeControls.js:679]: Yes it is removed in PR #239

File: src/Renderer/ThreeExtended/GlobeControls.js (hunk 21 above)
Comment (gchoqueux) [reply src/Renderer/ThreeExtended/GlobeControls.js]: Because the target must wait for the data to be loaded (in particular the relief data)

File: src/Renderer/c3DEngine.js (hunk 23 above)
Comment (gchoqueux) [reply src/Renderer/c3DEngine.js]: Yes, positionCamera and targetCamera are specific GlobeControl
There are removed in PR #239

File: src/Main.js [hunk 26]
```diff
@@ -4,13 +4,14 @@
...
 
-import ApiGlobe from './Core/Commander/Interfaces/ApiInterface/ApiGlobe';
+import ApiGlobe, { INITIALIZED_EVENT } from './Core/Commander/Interfaces/ApiInterface/ApiGlobe';
 // browser execution or not ?
 const scope = typeof window !== 'undefined' ? window : {};
 const itowns = scope.itowns || {
     viewer: new ApiGlobe(),
 };
 scope.itowns = itowns;
+itowns.viewer.INITIALIZED_EVENT = INITIALIZED_EVENT;
```
Comment (gchoqueux) [inline src/Main.js]: @autra how export INITIALIZED_EVENT?

**Comment by gchoqueux** [comment]
@autra PR updated

File: src/Main.js (hunk 26 above)
Comment (autra) [reply src/Main.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export

Comment (autra) [reply src/Main.js]: And you want to have a look at Main.js

**Comment by gchoqueux** [comment]
@autra PR updated

File: src/Renderer/c3DEngine.js (hunk 23 above)
Comment (autra) [reply src/Renderer/c3DEngine.js]: Where?

**Comment by autra** [comment]
@gchoqueux still some unanswered comments (https://github.com/iTowns/itowns2/pull/273/files/eaea9f8672ed35e22a10deca40bb8bb1b3a5b813#diff-df61a67eee71572718ef066564a3a0af). Thanks.

File: src/Renderer/ThreeExtended/GlobeControls.js (hunk 19 above)
Comment (gchoqueux) [reply src/Renderer/ThreeExtended/GlobeControls.js]: if there are  targets helper, we hide targets helper before picking
and after picking if there are  targets helper we display targets helper for rendering

Comment (gchoqueux) [reply src/Renderer/ThreeExtended/GlobeControls.js]: The use of THREE.js layers in PR @239 will simplify this point

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 27]
```diff
@@ -322,13 +367,9 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
...
     viewerDiv.addEventListener('globe-built', () => {
-        if (sceneIsLoaded === false) {
+        if (!sceneIsLoaded) {
             sceneIsLoaded = true;
-            this.scene.currentControls().updateCameraTransformation();
-            this.scene.updateScene3D();
-            viewerDiv.dispatchEvent(eventLoaded);
-        } else {
-            viewerDiv.dispatchEvent(eventError);
+            viewerDiv.dispatchEvent(new CustomEvent(DATA_LOADED_EVENT));
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: The event DATA_LOADED_EVENT is only listened to in this function. So you can remove it and put the 2 calls in the listener here. We don't need to add an event.

File: src/Renderer/ThreeExtended/GlobeControls.js [hunk 28]
```diff
@@ -1122,13 +1131,15 @@ function GlobeControls(camera, domElement, engine) {
...
     // update object camera position
-    this.updateCameraTransformation = function updateCameraTransformation(controlState)
+    this.updateCameraTransformation = function updateCameraTransformation(controlState, preventUpdate)
     {
         const bkDamping = this.enableDamping;
         this.enableDamping = false;
         state = controlState || CONTROL_STATE.ORBIT;
         update();
-        updateCameraTargetOnGlobe.bind(this)();
+        if (preventUpdate === undefined || preventUpdate) {
```
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js]: If preventUpdate is true, `updateCameraTargetOnGlobe` is called, which seems contradictory to me.

Did you want to 
```js
if (!preventUpdate) { // 
   // update
}
```

Or even better, do you know javascript [now has default parameters value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters).

I then suggest to have the following signature: 
```js
function updateCameraTransformation(controlState, updateCameraTarget = true)
```

 and simply use `if (updateCameraTarget)` here.

File: src/Renderer/ThreeExtended/GlobeControls.js [hunk 29]
```diff
@@ -478,18 +479,22 @@ function GlobeControls(camera, domElement, engine) {
...
+        const position = engine.getPickingPositionFromDepth(coords);
+        engine.renderScene();
 
-            return position;
-        };
-    }());
+        if (enableTargetHelper) {
+            pickingHelper.visible = true;
+            cameraTargetOnGlobe.visible = true;
+        }
```
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js:494]: Can you elaborate on these 2 ifs? I see no difference when I remove them. The `targetHelpers` are the axis on the globe where you clicked and where the target is, right? Picking is done with s+click right?

**Review by autra — CHANGES_REQUESTED** [review]
2 last remarks, one question, and we're good to go.

Before I merge, please split your review commits, and squash them in the relevant commits.

File: src/Renderer/ThreeExtended/GlobeControls.js (hunk 29 above)
Comment (gchoqueux) [reply src/Renderer/ThreeExtended/GlobeControls.js:494]: if 'enableTargetHelper' is enable there's error in picking position because helpers write in depth buffer. Picking position is do every time you click on globe.  s+click right is for selection.

> The targetHelpers are the axis on the globe where you clicked and where the target is, right

yes

**Comment by gchoqueux** [comment]
updated

**Comment by autra** [comment]
LGTM. A few nitpicks on commit messages with proposals:

>  fix (Camera) : fix setRange() doesn't return Promise
> 
>  return Promise in setRange()

Change to:
> Fix (Camera): Return a promise in setRange

In third commit:

> hide helperViewer when pick position

to 
> hide helperViewer when picking position to avoid wrong depth testing

In fifth:

> fix (GlobeControls) : fix initial target position is incorrect

to 
> Fix (GlobeControls): Correct initial target position

In sixth: there is a obsolete explanation about data-loaded (removed), and change:

> Fix (Camera): fix update target position after setRange()

to 
> Fix (Camera): fix target position update after setRange()

(if I understand well what you meant)

**Comment by gchoqueux** [comment]
commits message updated

**Review by autra — APPROVED** [review]

**Comment by autra** [comment]
Thanks!
